### ISSUE TITLE
Soundness of hashing writable sections

### ISSUE THREAD

**Issue by schuetzm** [leofidus/process_consistency#1]
I'm pretty sure that the current handling of writable sections is unsound. [In `compute_hash()`](https://github.com/leofidus/process_consistency/blob/a08aad7daeb2f88de89b67d6d2cc0663503082cf/src/lib.rs#L151), you're creating a slice to the section's memory. It has to be assumed that there exist (at least temporarily) mutable references to objects in that section (why would it be writable otherwise?), therefore creating any other aliasing reference to it is instant undefined behaviour. Even if not, any concurrent writes to the section while it is being read are a data race and therefore undefined behaviour as well.

I was going to suggest using volatile reads instead, however the [documentation of `read_volatile()`](https://doc.rust-lang.org/std/ptr/fn.read_volatile.html#safety) explicitly states:

> In particular, a race between a read_volatile and any write operation to the same location is undefined behavior.

... so this isn't a workaround either.

**Comment by leofidus** [comment]
Hm, that's a good point. I wouldn't expect anything bad to actually happen, but with undefined behavior who knows.

I don't see an easy workaround, other than removing the option (after all, hashing writable memory is a bit stupid in the first place, it's mostly there as an easy fix for cases where someone makes a page writable and then modifies it).

One version that might work in safe rust is to read from /proc/self/mem on Linux, or ReadProcessMemory on Windows. Both are more or less a copy of a snapshot of the memory, which avoids any aliasing; and in terms of concurrent access "the write happened when the kernel thinks so" isn't the worst model. The extra copy doesn't sound great from a performance perspective though.

**Comment by schuetzm** [comment]
Right, then it's disguising the memory access as IO, which would be fine. It might not be necessary, though: [`std::ptr::read()`](https://doc.rust-lang.org/std/ptr/fn.read.html) could work here. It lists only a few conditions for when it's UB; a data race or aliasing isn't one of them. OTOH it doesn't specifically say that it's safe either.

**Comment by leofidus** [comment]
It's just not just explicitly repeated. The documentation for [`std::ptr`](https://doc.rust-lang.org/std/ptr/index.html) still states 

> This means it is undefined behavior to perform two concurrent accesses to the same location from different threads unless both accesses only read from memory

The only notable improvement over the safety conditions on [std::slicefrom_raw_parts](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) seem to be that std::ptr::read/copy has a smaller window of unsoundness (copying is faster than hashing).

Doing some benchmarking with a medium-sized binary, using /proc/self/mem or ReadProcessMemory is about 30%-40% slower than the current approach. That's honestly not too bad. That approach also has proper error handling, and thus would resolve the documented unsoundness if any shared library is ever unloaded (which _will_ crash your program if you hit the race condition).

I'll change it to use the I/O approach by default, with the current fast-but-unsound direct memory access available behind an unsafe config option (since it can be sound if you uphold a list of invariants).
